### PR TITLE
[exporter/awsxray] Feature Gate to Allow fully random Trace IDs in XRay Exporter 

### PR DESCRIPTION
**Pull request by jj22ee**
**Description:** <Describe what has changed.>

Currently, the AWS XRay Exporter will drop segments/traces with trace IDs where the first 32 bits, when converted to UNIX Epoch time, are not within the past 28 days. This change is to add an Alpha feature gate `exporter.awsxray.skiptimestampvalidation`, which is disabled by default. If enabled,  the timestamp restriction is removed so that users do not need to use the AWS XRay ID Generator.

**Link to tracking Issue:** <Issue number if applicable>
N/A

**Testing:** <Describe what testing was performed and which tests were added.>
- Updated unit test in `config_test.go` 
- Added unit tests in `factory_test.go` and `segment_test.go`
- All other unit tests related to trace ID are passing.

**Documentation:** <Describe the documentation added.>
N/A

### CODE REVIEW

**Comment by atshaw43** [comment]
LGTM

File: exporter/awsxrayexporter/README.md [hunk 1]
```diff
@@ -26,12 +26,7 @@ propagated by them using the `X-Amzn-Trace-Id` HTTP header. However, other gener
 supported by replacing fully-random Trace IDs with X-Ray formatted Trace IDs where necessary:
 
 > AWS X-Ray IDs are the same size as W3C Trace Context IDs but differ in that the first 32 bits of a Trace ID
-> is the Unix epoch time when the trace was started. Since X-Ray only allows submission of Trace IDs from the
-> past 30 days, received Trace IDs are checked and spans without a valid timestamp are dropped.
-
-This means in order for spans to appear in X-Ray, the client SDK MUST use an X-Ray ID generator. For more
-information, see
-[configuring the X-Ray exporter](https://aws-otel.github.io/docs/getting-started/x-ray#configuring-the-aws-x-ray-exporter).
+> is the Unix epoch time when the trace was started.
```
Comment (pingleig) [inline exporter/awsxrayexporter/README.md:29]: Do we want to mention although the collector stop validating timestamp in trace id, X-Ray API will continue to validate it until it supports trace id with random number in the timestamp part of id. Meaning customer still need to use X-Ray id generator until service side support launches.

Comment (jj22ee) [reply exporter/awsxrayexporter/README.md:29]: Good point, for now I can mention that ID generator is still necessary. In the future, we will need to update this README once we actually don't need it.

File: exporter/awsxrayexporter/README.md [hunk 2]
```diff
@@ -26,8 +26,8 @@ propagated by them using the `X-Amzn-Trace-Id` HTTP header. However, other gener
 supported by replacing fully-random Trace IDs with X-Ray formatted Trace IDs where necessary:
 
 > AWS X-Ray IDs are the same size as W3C Trace Context IDs but differ in that the first 32 bits of a Trace ID
-> is the Unix epoch time when the trace was started. Since X-Ray only allows submission of Trace IDs from the
-> past 30 days, received Trace IDs are checked and spans without a valid timestamp are dropped.
+> is the Unix epoch time when the trace was started. Note that X-Ray only allows submission of Trace IDs from
+> the past 30 days, while the Exporter will not validate this timestamp.
```
Comment (bryan-aguilar) [inline exporter/awsxrayexporter/README.md:30]: What happens when a trace id is that is not valid is sent to the x-ray back end? What error will users see?

Comment (pingleig) [reply exporter/awsxrayexporter/README.md:30]: They will get 200 with all the ids in the unprocessed section https://docs.aws.amazon.com/xray/latest/api/API_PutTraceSegments.html#API_PutTraceSegments_ResponseElements

**Comment by jj22ee** [comment]
Updated PR - Move `remove timestamp validation` logic behind an Alpha feature gate.

File: exporter/awsxrayexporter/README.md (hunk 1 above)
Comment (jj22ee) [reply exporter/awsxrayexporter/README.md]: Change is removed in latest commit. This change will be added once the new `remove timestamp validation` logic is enabled by default

File: exporter/awsxrayexporter/internal/translator/segment.go [hunk 3]
```diff
@@ -22,6 +23,11 @@ import (
 	"github.com/open-telemetry/opentelemetry-collector-contrib/internal/coreinternal/traceutil"
 )
 
+var allowRandomTimestamp = featuregate.GlobalRegistry().MustRegister(
+	"exporter.awsxray.allowrandomtimestamp",
+	featuregate.StageAlpha,
+	featuregate.WithRegisterDescription("Remove XRay's timestamp validation on first 32 bits of trace ID"))
```
Comment (bryan-aguilar) [inline exporter/awsxrayexporter/internal/translator/segment.go:29]: I think there is a `featuregate.FromVersion()` option. Can you please add that with collector version `v0.84.0`.

**Review by bryan-aguilar — COMMENTED** [review]
Can you please add unit tests for the expected behavior after enabling the feature gate.

File: exporter/awsxrayexporter/internal/translator/segment.go [hunk 4]
```diff
@@ -22,6 +23,11 @@ import (
 	"github.com/open-telemetry/opentelemetry-collector-contrib/internal/coreinternal/traceutil"
 )
 
+var allowRandomTimestamp = featuregate.GlobalRegistry().MustRegister(
```
Comment (pingleig) [inline exporter/awsxrayexporter/internal/translator/segment.go:26]: nit: maybe change to skipTimestampValidation, we are not allowing random timestamp, we are allowing any random number with same length, it doesn't need to be a valid timestamp like 16xxxxx, 0000...01 also work.

File: exporter/awsxrayexporter/internal/translator/segment.go (hunk 3 above)
Comment (jj22ee) [reply exporter/awsxrayexporter/internal/translator/segment.go]: Added tests and `featuregate.WithRegisterFromVersion("v0.84.0")`

File: exporter/awsxrayexporter/internal/translator/segment.go (hunk 4 above)
Comment (jj22ee) [reply exporter/awsxrayexporter/internal/translator/segment.go]: Added suggestion

**Review by Aneurysm9 — APPROVED** [review]

File: exporter/awsxrayexporter/internal/translator/segment_test.go [hunk 5]
```diff
@@ -344,10 +344,43 @@ func TestSpanWithExpiredTraceId(t *testing.T) {
...
+
+func TestSpanWithExpiredTraceIdWithoutTimestampValidation(t *testing.T) {
+	// First Build expired TraceId
+	const maxAge = 60 * 60 * 24 * 30
+	ExpiredEpoch := time.Now().Unix() - maxAge - 1
+
+	tempTraceID := newTraceID()
+	binary.BigEndian.PutUint32(tempTraceID[0:4], uint32(ExpiredEpoch))
+
+	_, err := convertToAmazonTraceID(tempTraceID, true)
```
Comment (bryan-aguilar) [inline exporter/awsxrayexporter/internal/translator/segment_test.go:380]: Should we have assertions that the trace id is correct?

File: exporter/awsxrayexporter/internal/translator/segment_test.go [hunk 6]
```diff
@@ -344,10 +344,43 @@ func TestSpanWithExpiredTraceId(t *testing.T) {
...
+	attributes[conventions.AttributeHTTPTarget] = spanName
+	resource := constructDefaultResource()
+	span := constructClientSpan(pcommon.NewSpanIDEmpty(), spanName, ptrace.StatusCodeUnset, "OK", attributes)
+	timeEvents := constructTimedEventsWithSentMessageEvent(span.StartTimestamp())
+	timeEvents.CopyTo(span.Events())
+	traceID := span.TraceID()
+	traceID[0] = 0x11
+	span.SetTraceID(traceID)
+
+	_, err := MakeSegmentDocumentString(span, resource, nil, false, nil, true)
```
Comment (bryan-aguilar) [inline exporter/awsxrayexporter/internal/translator/segment_test.go:367]: Should there be assertions on this returned value?

File: exporter/awsxrayexporter/internal/translator/segment_test.go [hunk 7]
```diff
@@ -379,7 +412,7 @@ func TestServerSpanWithNilAttributes(t *testing.T) {
 	timeEvents.CopyTo(span.Events())
 	pcommon.NewMap().CopyTo(span.Attributes())
 
-	segment, _ := MakeSegment(span, resource, nil, false, nil)
+	segment, _ := MakeSegment(span, resource, nil, false, nil, false)
```
Comment (bryan-aguilar) [inline exporter/awsxrayexporter/internal/translator/segment_test.go:415]: Is an error being ignored here? Is there an assertion missing?

**Review by bryan-aguilar — APPROVED** [review]
The only nit I have is that there are a lot of errors being ignored in the tests.

File: exporter/awsxrayexporter/internal/translator/segment_test.go (hunk 6 above)
Comment (jj22ee) [reply exporter/awsxrayexporter/internal/translator/segment_test.go:367]: Assertions are in other test cases such as the `TestClientSpanWith*` tests. The scope of this test is to validate if an error doesn't occur.

**Review by srprash — APPROVED** [review]

File: exporter/awsxrayexporter/internal/translator/segment_test.go (hunk 6 above)
Comment (jj22ee) [reply exporter/awsxrayexporter/internal/translator/segment_test.go:367]: Nevermind, I will add a test here

File: exporter/awsxrayexporter/internal/translator/segment_test.go (hunk 7 above)
Comment (jj22ee) [reply exporter/awsxrayexporter/internal/translator/segment_test.go:415]: Looking at these tests, the assumption is that if there is a non-Nil segment (which is asserted), then the error should be nil (which is true looking at the return values of `MakeSegment`). These tests can be written better in a future change.

File: exporter/awsxrayexporter/internal/translator/segment_test.go (hunk 5 above)
Comment (jj22ee) [reply exporter/awsxrayexporter/internal/translator/segment_test.go]: Added assertion

**Comment by jj22ee** [comment]
@evan-bradley pinging for review and merge! We want to get this PR merged in time for the `v0.84.0` release.

**Comment by jj22ee** [comment]
The failing tests should be unrelated to this PR.

**Pull request merged**